Simplify test utxo freezing

hewison-chris opened the pull request:
Created a `base.d` function `freezeUTXO` which uses the new `getSpendables` that finds unspent `utxo` from the `utxo_set` of a node. Then replaced some of the complex code in the test cases by using `freezeUTXO`. 
In some cases the test could then also use `generateBlocks` to further simplify.

hewison-chris committed:
Tests: Simplify `test.EnrollDifferentUTXO`

Use `generateBlocks`, `postAndEnsureTxInPool`, `getSpendables` and
`freezeUTXO` from base class.

hewison-chris committed:
Tests: Add `freezeUTXO` to base test class

This will generate frozen utxo by using the new `getSpendables` on the
first client to fetch unspent utxo from the utxo_set of that client. The
frozen utxo are output to each of the nodes indicated by indexes
provided so that they will be able to enroll in a future block.

hewison-chris committed:
Tests: Simplify `test.FutureEnrollment`

Use `postAndEnsureTxInPool` and `freezeUTXO` from base class.

hewison-chris committed:
Tests: Simplify `test.Quorum`

Use `postAndEnsureTxInPool` and `freezeUTXO` from base class.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -1092,6 +1093,37 @@ public class TestAPIManager
+        assert(expected.mul(keys.length));
+        auto utxo_pairs = first_client.getSpendables(expected, OutputType.Payment);
+
+        return TxBuilder(WK.Keys.AAA.address) // Refund
```
This is hard code, I think. The `freezeUtxo` could have a refund address as a parameter. It would be better to know where the refund goes without looking into this function.

hewison-chris replied:
I'll add with a default to keep the calling code more concise.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -1092,6 +1093,37 @@ public class TestAPIManager
+
+    ***************************************************************************/
+
+    Transaction freezeUtxo (Idxs)(Idxs client_idxs)
```
```suggestion
    Transaction freezeUTXO (Idxs)(Idxs client_idxs)
```

linked0 reviewed (commented):
Nice but I have some suggestions.

hewison-chris committed:
Tests: Simplify `test.RestoreSlashingInfo`

Use `freezeUTXO`, `postAndEnsureTxInPool` and `getSpendables` of base
test class.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1092,6 +1093,39 @@ public class TestAPIManager
+    {
+        static assert (isInputRange!Idxs);
+
+        auto first_client = this.clients[0];
+
+        const keys = client_idxs.map!(i => this.nodes[i].getPublicKey().key).array;
```
Why not provide a range of public key instead ? This looks like a lot of coupling for little convenience.

hewison-chris replied:
This keeps it in line with all the other functions in this base file though. They take a range of indexes. Do we want this to be special case?

Geod24 reviewed (approved):
Just one thing

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1092,6 +1093,39 @@ public class TestAPIManager
+        assert(expected.mul(keys.length));
+        auto utxo_pairs = first_client.getSpendables(expected, OutputType.Payment);
+
+        return TxBuilder(refund_me)
```
Or just use `utxo_pairs[0].utxo.output.address` ?

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1092,6 +1093,39 @@ public class TestAPIManager
+        // Collect enough utxo for all to enroll
+        Amount expected = Amount.MinFreezeAmount;
+        assert(expected.mul(keys.length));
+        auto utxo_pairs = first_client.getSpendables(expected, OutputType.Payment);
```
Extra param

hewison-chris replied:
Remember that the default param did not work for localRest API.

kchulj reviewed (approved):
LGTM
